ur_controllers/src/tool_contact_controller.cpp

```diff
@@ -0,0 +1,384 @@
+    auto result = std::make_shared<ur_msgs::action::ToolContact::Result>();
+    active_goal->setAborted(result);
+    should_reset_goal = true;
+  }
+  if (tool_contact_active_) {
+    tool_contact_active_ = false;
+    if (!tool_contact_set_state_interface_->get().set_value(TOOL_CONTACT_WAITING_END)) {
+      RCLCPP_FATAL(get_node()->get_logger(), "Controller failed to update command interface.");
+      return controller_interface::CallbackReturn::ERROR;
+    }
+  }
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn ToolContactController::on_shutdown(const rclcpp_lifecycle::State&
+                                                                        /* previous_state */)
+{
+  tool_contact_action_server_.reset();
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_action::GoalResponse ToolContactController::goal_received_callback(
+    const rclcpp_action::GoalUUID& /*uuid*/, std::shared_ptr<const ur_msgs::action::ToolContact::Goal> /* goal */)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "New goal received.");
+
+  if (get_lifecycle_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    RCLCPP_ERROR(get_node()->get_logger(), "Tool contact controller is not in active state, can not accept action "
+                                           "goals.");
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+
+  const auto active_goal = *rt_active_goal_.readFromNonRT();
+  if (active_goal) {
+    RCLCPP_ERROR(get_node()->get_logger(), "Tool contact already active, rejecting goal.");
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+void ToolContactController::goal_accepted_callback(
+    std::shared_ptr<rclcpp_action::ServerGoalHandle<ur_msgs::action::ToolContact>> goal_handle)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Goal accepted.");
+  tool_contact_enable_ = true;
+  tool_contact_abort_ = false;
+  RealtimeGoalHandlePtr rt_goal = std::make_shared<RealtimeGoalHandle>(goal_handle);
+  rt_goal->execute();
+  rt_active_goal_.writeFromNonRT(rt_goal);
+  goal_handle_timer_.reset();
+  auto period = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      std::chrono::nanoseconds(action_monitor_period_.nanoseconds()));
+  goal_handle_timer_ = get_node()->create_wall_timer(period, std::bind(&ToolContactController::action_handler, this));
+  return;
+}
+
+void ToolContactController::action_handler()
+{
+  const auto active_goal = *rt_active_goal_.readFromNonRT();
+  if (active_goal) {
+    // Allow the goal to handle any actions it needs to perform
+    active_goal->runNonRealtime();
+    // If one of the goal ending conditions were met, reset our active goal pointer
+    if (should_reset_goal) {
+      rt_active_goal_.writeFromNonRT(RealtimeGoalHandlePtr());
+      should_reset_goal = false;
+    }
+  }
+}
+
+rclcpp_action::CancelResponse ToolContactController::goal_cancelled_callback(
```

Comment:
```suggestion
rclcpp_action::CancelResponse ToolContactController::goal_canceled_callback(
```

We want use American English. I know, we haven't been doing that everywhere, but let's try to stick to that in the scope of this PR.